craft_application/models/spread.py

```diff
@@ -0,0 +1,242 @@
+# This file is part of craft-application.
+#
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Models representing spread projects."""
+
+
+from typing_extensions import Any, Self
+
+from craft_application.models import CraftBaseModel
+
+# Simplified spread configuration
+
+
+class SimpleSpreadSystem(CraftBaseModel):
+    """Simplified spread system configuration."""
+
+    workers: int | None = None
+
+
+class SimpleSpreadBackend(CraftBaseModel):
+    """Simplified spread backend configuration."""
+
+    type: str
+    allocate: str | None = None
+    discard: str | None = None
+    systems: list[dict[str, SimpleSpreadSystem | None]]
+    prepare: str | None = None
+    restore: str | None = None
+    prepare_each: str | None = None
+    restore_each: str | None = None
+
+
+class SimpleSpreadSuite(CraftBaseModel):
+    """Simplified spread suite configuration."""
+
+    summary: str
+    systems: list[str]
+    environment: dict[str, str] | None = None
+    prepare: str | None = None
+    restore: str | None = None
+    prepare_each: str | None = None
+    restore_each: str | None = None
+    kill_timeout: str | None = None
+
+
+class SimpleSpreadYaml(CraftBaseModel):
+    """Simplified spread project configuration."""
+
+    backends: dict[str, SimpleSpreadBackend]
+    suites: dict[str, SimpleSpreadSuite]
+    exclude: list[str] | None = None
+    prepare: str | None = None
+    restore: str | None = None
+    prepare_each: str | None = None
+    restore_each: str | None = None
+    kill_timeout: str | None = None
+
+
+# Processed full-form spread configuration
+
+
+class SpreadBaseModel(CraftBaseModel):
+    """Base for spread models."""
+
+    def model_post_init(self, __context: Any) -> None:  # noqa: ANN401
+        """Remove attributes set to None."""
+        none_items: list[Any] = []
+        for k, v in self.__dict__.items():
+            if v is None:
+                none_items.append(k)
+
+        for k in none_items:
+            k.replace("_", "-")
+            delattr(self, k)
```

Comment:
I had some problems with `exclude_unset` it but I will re-investigate it and use it instead if possible.

Reply:
The base class already sets `exclude_unset`. The model has the field set to `None`, making it appear in the dumped file. The `kill_timeout` case still works because field exclusion is applied to the final model, not the simple one.